src/v/v8_engine/executor.h

```diff
@@ -0,0 +1,190 @@
+
+    ss::future<> start();
+    ss::future<> stop();
+
+    // Try to acquire semaphore and sleep if semaphore does not contain free
+    // slots
+    ss::future<> acquire();
+
+    // Release sempahore
+    void release();
+
+private:
+    const size_t max_free_slots;
+    std::atomic<size_t> _free_slots;
+
+    static constexpr ss::lowres_clock::duration _max_wait_time_ms{100};
+
+    ss::sharded<ss::gate> _gates;
+
+    // _abort_sources is used for abort sleep if semaphore is stoped
+    ss::sharded<ss::abort_source> _abort_sources;
+
+    std::atomic<bool> _is_closed{false};
+};
+
+} // namespace internal
+
+// This class implement thread pool with std::thread for runing v8 script. It
+// uses boost::lockfree_queue for storing task. Queue task stores #thread_count
+// tasks. Because we need to set watchdog for timeout v8 script only when
+// executor can execute script without wating. semaphore is used for controle
+// does executor have free thread.
+class executor {
+public:
+    executor(size_t threads_num, uint8_t cpu_id);
+
+    executor(const executor& other) = delete;
+    executor& operator=(const executor& other) = delete;
+    executor(executor&& other) = delete;
+    executor& operator=(executor&& other) = delete;
+
+    ~executor();
+
+    // Start ss::sharded for ss::gate and ss::abort_source
+    ss::future<> start();
+
+    // Stop executor. Frst of all we need to stop semaphore for cancel waitnig
+    // tasks. After that we need complete all task in queue and finally stop
+    // std::threads
+    ss::future<> stop();
+
+    bool is_stopping() const;
+
+    /// Submit new task in executor.
+    ///
+    /// \param start_func is used for set watchdog
+    /// \param func_for_executor is v8 script
+    template<typename StartFunc, typename FuncForExecutor>
+    auto submit(StartFunc&& start_func, FuncForExecutor&& func_for_executor) {
+        return _sem.acquire()
+          .then([this,
+                 start_func = std::forward<StartFunc>(start_func),
+                 func_for_executor = std::forward<FuncForExecutor>(
+                   func_for_executor)]() mutable {
+              auto new_task = std::make_unique<internal::task<FuncForExecutor>>(
+                std::forward<FuncForExecutor>(func_for_executor));
+              auto fut = new_task->get_future();
+              _tasks.push(new_task.get());
+              _has_task_cv.notify_one();
+              start_func(); // Rearm watchdog for srcipt
+              return fut.finally([new_task = std::move(new_task)] {});
+          })
+          .finally([this] { _sem.release(); });
+    }
+
+private:
+    // Do we need to pin all thread for one core?
+    void pin(unsigned cpu_id);
+
+    // Main loop for threads in executor
+    void loop();
+
+    std::atomic<bool> _is_stopped{false};
+
+    std::mutex mutex;
+    std::condition_variable _has_task_cv;
+    std::vector<std::thread> _threads; // maybe use ss::alien_thread
```

Comment:
yes, absolutely. this would violate seastar's primitives like CPU scheduling, etc and starve the main thread. it would also break the concurrency model in seastar.